docs/architecture/psa-thread-safety.md

```diff
@@ -67,16 +67,28 @@ In the medium to long term, performing a slow or blocking operation (for example
 
 We may want to go directly to a more sophisticated approach because when a system works with a global lock, it's typically hard to get rid of it to get more fine-grained concurrency.
 
+### Key destruction short-term requirements
+
+#### Summary of guarantees when `psa_destroy_key` returns
+
+* The key identifier doesn't exist. Rationale: this is a functional requirement for persistent keys: the caller can immediately create a new key with the same identifier.
+* The resources from the key have been freed. Rationale: in a low-resource condition, this may be necessary for the caller to re-create a similar key, which should be possible.
+* The call must not block indefinitely, and in particular cannot wait for an event that is triggered by application code such as calling an abort function. Rationale: this may not strictly be a functional requirement, but it is an expectation `psa_destroy_key` does not block forever due to another thread, which could potentially be another process on a multi-process system.
+
+As noted above in [“Correctness out of the box”](#correctness-out-of-the-box), destroying a key while it's in use is undefined behavior. The only guarantee in this case is that it won't cause data corruption or read-after-free inside the key store. In particular, any or all the listed guarantees may be violated when `psa_destroy_key` is called on a key that is in use.
```

Comment:
> For example, a race condition causing data from a destroyed key to leak into the next key that uses the same slot would be a critical security flaw

Arguably, this could be considered data corruption. But I see what you mean, I'll try to make it more general if this sentence remains even after we discussed the remaining points.

Reply:
I phrased it this way because this seemed to be the minimum for compliance and functional correctness. Ideally we wouldn't violate them, but it gives us more space in design and to do incremental development.

> The requirement on resources is somewhat flexible because apart from the key slot itself, resources can be considered to be associated with the ongoing operation rather than the key. I don't think there's any advantage to be gained anyway by keeping the key slot busy if psa_destroy_key is called while an operation is active, but not while an operation is idle.

The main reason I wrote this clause about undefined behaviour is to allow us violate this property. I wanted requirements that allow a relatively simple solution that could be incrementally improved. I don't know how to keep this property when there are ongoing operations without condition variables or uuid.